### ISSUE TITLE
VS 2019 cpp/custom-dataset example compiler error

### ISSUE THREAD

**Issue by alex-zador** [pytorch/pytorch#35123]
When compiling custom-dataset example receive errors:
1) no instance of constructor "torch::nn::Functional::Functional" matches the argument list	CustomDataSet	...\CustomDataSet\CTorch.cpp	125	
2) cannot determine which instance of overloaded function "torch::log_softmax" is intended	CustomDataSet	...\CustomDataSet\CTorch.cpp	125	
 line 125 is: push_back(Functional(torch::log_softmax, 1, torch::nullopt));

Could anybody help with this?

cc @yf225 @peterjc123

**Comment by krshrimali** [comment]
Thanks for creating this issue.

A temporary work around is using `torch::nn::functional::log_softmax` instead of `torch::log_softmax` in `custom-dataset.cpp`. 

I'm not sure if this is the desired change, as per the recent changes in libtorch API. After a few discussions, I'll create a PR to change this in examples by the weekend. Thanks!

**Comment by alex-zador** [comment]
Actually, I've tried this but receive another error for the same line. 

Compile succeeded when I used 
push_back(Functional(torch::nn::functional::detail::log_softmax, 1, torch::nullopt));

I will see if the code work at runtime...

Should I close this issue?

**Comment by krshrimali** [comment]
Sorry, I wasn't clear about this change. Using `torch::nn::functional::log_softmax` requires to remove the third argument (`torch::nullopt`). Using `detail` namespace works but I doubt if it's supposed to be user-facing namespace.

Please keep this issue open (until we hear from @yf225 on this) and let us know if the code works well at runtime.

**Comment by alex-zador** [comment]
Th executable ends with c10::IndexError in KernelFunction.h

**Comment by alex-zador** [comment]
Th executable ends up with c10::IndexError in KernelFunction.h at first train iteration.
It runs in CPU mode.

I am using VS2019, OpenCV 4.2.

**Comment by krshrimali** [comment]
Thank you for the information @alex-zador. I need a little more information on this:

1. Are you using Caltech-101 Data Set mentioned here: https://github.com/pytorch/examples/tree/master/cpp/custom-dataset?
2. Can you copy the whole error here? It will be helpful to debug.

Also, please re-open this issue since the issue is not technically resolved. I'll work on the cpp-customdataset example this weekend, thanks!

**Comment by alex-zador** [comment]
@krshrimali, sorry, that was unintentionally closed.

**Comment by alex-zador** [comment]
@krshrimali, sure I used  Caltech-101 and info.txt.
 
That is strange, but this time code runned a number of iterations. I received:
Running on: CPU
Train Epoch: 1 16/7281  Loss: 0.297678  Acc: 0
Train Epoch: 1 176/7281 Loss: 0.523969  Acc: 0.0965909
Train Epoch: 1 336/7281 Loss: 0.530743  Acc: 0.0922619
Train Epoch: 1 496/7281 Loss: 0.53766   Acc: 0.0927419
Train Epoch: 1 656/7281 Loss: 0.536154  Acc: 0.0990854
Train Epoch: 1 816/7281 Loss: 0.535026  Acc: 0.10049
Train Epoch: 1 976/7281 Loss: 0.537718  Acc: 0.0942623

C:\Users\alex\...\x64\Release\CustomDataSet.exe (process 13020) exited with code -1073740791.

Next time that was:
Running on: CPU
Train Epoch: 1 16/7281  Loss: 0.297678  Acc: 0

C:\Users\alex\...\x64\Release\CustomDataSet.exe  (process 1980) exited with code -1073740791.

and then:
Running on: CPU

C:\Users\alex\...\x64\Release\CustomDataSet.exe  (process 14504) exited with code -1073740791.

If I run it as Debug (F5) that breaks with 
![CDS Error](https://user-images.githubusercontent.com/39754219/77244262-27c3fe80-6c24-11ea-8cf6-31d0589a96c2.png)

and 

Exception thrown at 0x00007FFE7B879179 in CustomDataSet.exe: Microsoft C++ exception: c10::IndexError at memory location 0x00000007A71DCA20.
Unhandled exception at 0x00007FFE7B879179 in CustomDataSet.exe: Microsoft C++ exception: c10::IndexError at memory location 0x00000007A71DCA20.

in Output window.

**Comment by krshrimali** [comment]
Thank you for sharing this. I'll run on my local system and see if I can reproduce this.

**Comment by krshrimali** [comment]
Update: Sorry, it took time. I was able to reproduce this issue with:

`VS 2019, C++ 14 ISO Version, Libtorch (PyTorch 1.4 Stable) Debug Version`

The build is successful but the following exception is thrown:

```
Exception thrown at 0x00007FFBCDB4A839 in custom-dataset.exe: Microsoft C++ exception: c10::IndexError at memory location 0x0000007E92CFAE80. occurred
```

Exception thrown at:

```
if (C10_LIKELY(unboxed_kernel_func_ != nullptr)) {
      using ActualSignature = Return (OperatorKernel*, Args...);
      ActualSignature* func = reinterpret_cast<ActualSignature*>(unboxed_kernel_func_);
      return (*func)(getFunctor_(), std::forward<Args>(args)...); // --- here
}
```

Here is the **Call Stack**:

```
custom-dataset.exe!c10::KernelFunction::callUnboxedOnly<at::Tensor,at::Tensor const &,at::Tensor const &,at::Tensor const &,__int64,__int64>(const at::Tensor & <args_0>, const at::Tensor & <args_1>, const at::Tensor & <args_2>, __int64 <args_3>, __int64 <args_4>) Line 95	C++
custom-dataset.exe!c10::Dispatcher::doCallUnboxedOnly::__l2::<lambda>(const ska::flat_hash_map<enum c10::TensorTypeId,c10::KernelFunction,std::hash<enum c10::TensorTypeId>,std::equal_to<enum c10::TensorTypeId>,std::allocator<std::pair<enum c10::TensorTypeId,c10::KernelFunction>>> & backendFallbackKernels) Line 211	C++
custom-dataset.exe!c10::LeftRight<ska::flat_hash_map<enum c10::TensorTypeId,c10::KernelFunction,std::hash<enum c10::TensorTypeId>,std::equal_to<enum c10::TensorTypeId>,std::allocator<std::pair<enum c10::TensorTypeId,c10::KernelFunction>>>>::read<at::Tensor <lambda>(const ska::flat_hash_map<enum c10::TensorTypeId,c10::KernelFunction,std::hash<enum c10::TensorTypeId>,std::equal_to<enum c10::TensorTypeId>,std::allocator<std::pair<enum c10::TensorTypeId,c10::KernelFunction>>> &)>(c10::Dispatcher::doCallUnboxedOnly::__l2::at::Tensor <lambda>(const ska::flat_hash_map<enum c10::TensorTypeId,c10::KernelFunction,std::hash<enum c10::TensorTypeId>,std::equal_to<enum c10::TensorTypeId>,std::allocator<std::pair<enum c10::TensorTypeId,c10::KernelFunction>>> &) && readFunc) Line 74	C++
custom-dataset.exe!c10::Dispatcher::doCallUnboxedOnly<at::Tensor,at::Tensor const &,at::Tensor const &,at::Tensor const &,__int64,__int64>(const c10::DispatchTable & dispatchTable, const c10::LeftRight<ska::flat_hash_map<enum c10::TensorTypeId,c10::KernelFunction,std::hash<enum c10::TensorTypeId>,std::equal_to<enum c10::TensorTypeId>,std::allocator<std::pair<enum c10::TensorTypeId,c10::KernelFunction>>>> & backendFallbackKernels, const at::Tensor & <args_0>, const at::Tensor & <args_1>, const at::Tensor & <args_2>, __int64 <args_3>, __int64 <args_4>) Line 208	C++
custom-dataset.exe!c10::Dispatcher::callUnboxedOnly::__l2::<lambda>(const c10::DispatchTable & dispatchTable) Line 201	C++
custom-dataset.exe!c10::LeftRight<c10::DispatchTable>::read<at::Tensor <lambda>(const c10::DispatchTable &)>(c10::Dispatcher::callUnboxedOnly::__l2::at::Tensor <lambda>(const c10::DispatchTable &) && readFunc) Line 74	C++
custom-dataset.exe!c10::impl::OperatorEntry::readDispatchTable<at::Tensor <lambda>(const c10::DispatchTable &)>(c10::Dispatcher::callUnboxedOnly::__l2::at::Tensor <lambda>(const c10::DispatchTable &) && functor) Line 32	C++
custom-dataset.exe!c10::Dispatcher::callUnboxedOnly<at::Tensor,at::Tensor const &,at::Tensor const &,at::Tensor const &,__int64,__int64>(const c10::OperatorHandle & op, const at::Tensor & <args_0>, const at::Tensor & <args_1>, const at::Tensor & <args_2>, __int64 <args_3>, __int64 <args_4>) Line 198	C++
custom-dataset.exe!at::nll_loss(const at::Tensor & self, const at::Tensor & target, const at::Tensor & weight, __int64 reduction, __int64 ignore_index) Line 14270	C++
custom-dataset.exe!train<torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<CustomDataset,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::SequentialSampler>>(Network & network, torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<CustomDataset,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::SequentialSampler> & loader, torch::optim::Optimizer & optimizer, unsigned __int64 epoch, unsigned __int64 data_size) Line 148	C++
custom-dataset.exe!main() Line 223	C++
```

@peterjc123 - I understand you must be very busy, but could you look at this? MNIST Example works fine on the same configuration, but this fails. Is this anyhow related to https://github.com/pytorch/pytorch/pull/35362? You are more than welcome to ask for any other information or any help you need related to this issue, to better understand this problem.

**Note**: In `custom-dataset.cpp`, I tried printing labels just before `torch::nll_loss` call. And for the first batch of data, they turned out to be:

```
-x.xxxx10e18
```

It should have interpreted it as `0`. So I changed https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp#L58 to: `torch::Tensor tlabel = torch::full({1}, data[index].second);`.

**Comment by peterjc123** [comment]
Yeah, I'm able to reproduce this issue. And I manage to resolve it by replacing all the occurrences of `long` into `int64_t`.

**Comment by krshrimali** [comment]
Thank you for suggesting the resolve. I'm curious to know why using `int64_t` worked and `long` didn't?

**Comment by peterjc123** [comment]
`label` is declared as `kLong` in https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp#L58 but it is used as `long` in https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp#L73 and https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp#L23. `kLong` refers to `kInt64`, which is a 8-bit data type, but `long` in MSVC is a 4-bit data type. There may be some data loss during the implicit casting between those two data types.

**Comment by krshrimali** [comment]
Thank you for the explanation.

@alex-zador: You can close this issue after testing the solution suggested by @peterjc123 above (it works fine on my machine). Though, the custom-dataset example might need some refactoring, so don't wait for the PR (it may take some time). :)

**Comment by alex-zador** [comment]
Thank you guys. Now it works on my machine too.

**Comment by stalek71** [comment]
It looks the sample has not been updated up to now and still contains build errors.
Probably it should be reopened...

![obraz](https://user-images.githubusercontent.com/6651025/147592202-de83ba69-6f7f-4f5d-8aef-d5fd8acdaee9.png)
